File: LeetCode/901-1000/973. K Closest Points to Origin/src/Solution.java
import java.util.PriorityQueue;

class Solution {
    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> queue = new PriorityQueue<>();
        for (int[] point : points) {
            queue.add(new Point(point));
            if (queue.size() > k) {
                queue.poll();
            }
        }
        int[][] ans = new int[k][];
        for (int i = 0; i < k; i++) {
            ans[i] = queue.poll().cords;
        }
        return ans;
    }

    private static class Point implements Comparable<Point> {
        private int[] cords;
        private int length;

        private Point(int[] cords) {
            this.cords = cords;
            this.length = (cords[0] * cords[0]) + (cords[1] * cords[1]);
        }

        @Override
        public int compareTo(Point o) {
            return -Integer.compare(this.length, o.length);
        }
    }
}

File: LeetCode/901-1000/973. K Closest Points to Origin/Solution.py
import heapq as hq


class Solution:
    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        closest_points = []
        for point in points:
            dist = point[0] ** 2 + point[1] ** 2
            hq.heappush(closest_points, (-dist, point))
            if len(closest_points) > k:
                hq.heappop(closest_points)
        for i in range(len(closest_points)):
            closest_points[i] = closest_points[i][1]
        return closest_points